### PR TITLE
[CUBRIDMAN-13] manual for Hash List Scan and Optional Statistics Estimation

### PR DESCRIPTION
**Pull request by shparkcubrid**
http://jira.cubrid.org/browse/CUBRIDMAN-13

CBRD-23665 Support Hash List Scan
CBRD-23732 Optional Statistics Estimation

add 2 system parameter and 1 sql hint.

**system parameter**
max_hash_list_scan_size: Maximum temp file size for hash list scan.
```
default : 4MB, min : 0, max : 128MB
```
use_stat_estimation : whether to use estimation statistics
```
default : NO 
```

**SQL hint**
```
NO_HASH_LIST_SCAN: Not use hash list scan.
```

**related page**
https://www.cubrid.org/manual/en/10.2/sql/tuning.html#using-sql-hint
https://www.cubrid.org/manual/en/10.2/admin/config.html#system-parameters

### CODE REVIEW

File: en/admin/config.rst [hunk 1]
```diff
@@ -1992,6 +2008,10 @@ The following are other parameters. The type and value range for each parameter
 
     **optimizer_enable_merge_join** is a parameter to specify whether to include sort merge join plan as a candidate of query plans or not. The default is **no**. Regarding sort merge join, see :ref:`sql-hint`.
 
+**use_stat_estimation**
+
+    **use_stat_estimation** is a parameter to specify whether to use the estimated information in calculating statistics or not. The default is no. The estimated information generated by the heap manager during DML is related to the number of added objects. it is relatively accurate for number of total objects, NOT for number of distinct values.
+
```
Comment (beyondykk9) [inline en/admin/config.rst:2014]: during DML --> while processing DML
related to the number of added objects --> associated with the number of added objects
it is relatively accurate for number of total objects, NOT for number of distinct values.
-->
It is relatively accurate for the number of total objects, not for the number of distinct values.

File: en/sql/tuning.rst [hunk 2]
```diff
@@ -510,13 +530,13 @@ The following are the explanation regarding items of trace statistics.
 *   sort: sorting or not
 *   page: the number of pages which is read in this operation; the number of used pages except the internal sorting buffer
 *   rows: the number of the result rows in this operation
+*   hash: hash aggregate evaluation or not, when sorting tuples in the aggregate function(true/false). See :ref:`NO_HASH_AGGREGATE <no-hash-aggregate>` hint.
```
Comment (charischau) [inline en/sql/tuning.rst:533]: I m not sure whether this is what you want to express? 

hash: hash aggregate evaluation or not, when sorting tuples in the aggregate function(true/false).
-->
hash: whether to hash aggregate evaluation when sorting tuples in the aggregate function(true/false).

File: en/sql/tuning.rst [hunk 3]
```diff
@@ -498,6 +509,15 @@ The following are the explanation regarding items of trace statistics.
...
     *   rows: the number of result rows when this operation is performed; the number of result rows to which key filter is applied
+
+*   temp: data scanning job with temp file
+
+    *   hash: hash list scan or not. See :ref:`NO_HASH_LIST_SCAN <no-hash-list-scan>` hint.
+    *   buildtime: the estimated time(ms) in building hash table.
+    *   time: the estimated time(ms) in probing hash table.
+    *   fetch, ioread: page fetching count and I/O read count in the temp file of this operation
+    *   readrows: the number of read rows when this operation is performed
+    *   rows: the number of result rows when this operation is performed
```
Comment (beyondykk9) [inline en/sql/tuning.rst:520]: readrows: the number of  read rows --> readrows: the number of rows read

File: en/sql/tuning.rst [hunk 4]
```diff
@@ -649,6 +670,14 @@ The following hints can be specified in **UPDATE**, **DELETE** and **SELECT** st
     
         Hash aggregate evaluation will not work for functions evaluated on distinct values (e.g. AVG(DISTINCT x)) and for the GROUP_CONCAT and MEDIAN functions, since they require an extra sorting step for the tuples of each group.
 
+.. _no-hash-list-scan:
+
+*   **NO_HASH_LIST_SCAN**: This is a hint not to use hash list scan for scanning sub-query's result. Instead, list scan is used to scan temp file. By building and probing hash table, we can reduce the amount of data that needs to be searched. However, in some scenarios the user may know beforehand that outer cardinality is very small and can use the hint to skip hash list scan entirely. For setting the memory size of hash scan, see :ref:`max_hash_list_scan_size <max_hash_list_scan_size>`.
```
Comment (charischau) [inline en/sql/tuning.rst]: However, in some scenarios the user may know...
-->
However, in some scenarios, the user may know...

File: en/sql/tuning.rst [hunk 5]
```diff
@@ -649,6 +670,14 @@ The following hints can be specified in **UPDATE**, **DELETE** and **SELECT** st
...
+
+    .. note::
+    
+        Hash List scan only work for predicates having a equal operation and do NOT work for predicate having oid type.
```
Comment (charischau) [inline en/sql/tuning.rst]: Hash List scan only work for predicates having a equal operation and do NOT work for predicate having oid type.
 -->
Hash List scan only works for predicating having an equal operation and does NOT work for predicating having OID type.

File: en/admin/config.rst (hunk 1 above)
Comment (shparkcubrid) [reply en/admin/config.rst:2014]: I fixed it

File: en/sql/tuning.rst (hunk 4 above)
Comment (shparkcubrid) [reply en/sql/tuning.rst]: I fixed it.

File: en/sql/tuning.rst (hunk 3 above)
Comment (shparkcubrid) [reply en/sql/tuning.rst:520]: I fixed it.

File: en/sql/tuning.rst (hunk 2 above)
Comment (shparkcubrid) [reply en/sql/tuning.rst:533]: Actually I only changed the position of this sentence. I think existing sentence is enough understandable. could you check again?

File: en/sql/tuning.rst (hunk 5 above)
Comment (shparkcubrid) [reply en/sql/tuning.rst]: 'Hash List scan only works for predicates having an equal operation and does NOT work for predicates having OID type.'
I think this is better. the 'predicate' was written as noun.

File: en/sql/tuning.rst (hunk 2 above)
Comment (charischau) [reply en/sql/tuning.rst:533]: okay it makes sense

**Review by charischau — DISMISSED** [review]

File: en/admin/config.rst [hunk 6]
```diff
@@ -138,6 +138,8 @@ On the below table, if "Applied" is "server parameter", that parameter affects t
 |                               +-------------------------------------+-------------------------+---------+----------+--------------------------------+-----------------------+
 |                               | max_agg_hash_size                   | server parameter        |         | byte     | 2,097,152(2M)                  |                       |
 |                               +-------------------------------------+-------------------------+---------+----------+--------------------------------+-----------------------+
+|                               | max_hash_list_scan_size             | server parameter        |         | byte     | 4M                             |                       |
```
Comment (hgryoo) [inline en/admin/config.rst]:
```suggestion
|                               | max_hash_list_scan_size             | server parameter        |         | byte     | 4,194,304(4M)                  |                       |
```

File: en/admin/config.rst [hunk 7]
```diff
@@ -558,6 +562,8 @@ The following are parameters related to the memory used by the database server o
 +--------------------------------+--------+---------------------------+---------------------------+---------------------------+
 | max_agg_hash_size              | byte   | 2,097,152(2M)             | 32,768(32K)               | 134,217,728(128MB)        |
 +--------------------------------+--------+---------------------------+---------------------------+---------------------------+
+| max_hash_list_scan_size        | byte   | 4M                        | 0                         | 128MB                     |
```
Comment (hgryoo) [inline en/admin/config.rst]:
```suggestion
| max_agg_hash_size              | byte   | 2,097,152(2M)             | 32,768(32K)               | 134,217,728(128MB)        |
+--------------------------------+--------+---------------------------+---------------------------+---------------------------+
| max_hash_list_scan_size        | byte   | 4,194,304(4M)            | 0                         | 128MB                     |
```

**Review by hgryoo — COMMENTED** [review]
It'd better write the exact number as written like max_agg_hash_size

**Review by hgryoo — APPROVED** [review]

**Review by charischau — APPROVED** [review]

**Review by beyondykk9 — APPROVED** [review]

**Pull request merged**